File: Difficulty: Medium/Sort Elements by Decreasing Frequency/sort-elements-by-decreasing-frequency.cpp
class Solution {
public:
    vector<int> sortByFreq(vector<int>& arr) {
        unordered_map<int, int> freq;

        // Count frequency
        for (int num : arr) {
            freq[num]++;
        }

        // Sort using frequency and value
        sort(arr.begin(), arr.end(), [&](int a, int b) {
            if (freq[a] != freq[b]) {
                return freq[a] > freq[b];
            }

            return a < b;
        });

        return arr;
    }
};

File: Difficulty: Medium/Sort Elements by Decreasing Frequency/sort-elements-by-decreasing-frequency.java
class Solution {
    // Function to sort the array according to frequency of elements.
    public ArrayList<Integer> sortByFreq(int arr[]) {
        HashMap<Integer, Integer> map = new HashMap<>();

        for (int num : arr) {
            map.put(num, map.getOrDefault(num, 0) + 1);
        }

        ArrayList<Integer> result = new ArrayList<>();

        for (int num : arr) {
            result.add(num);
        }

        result.sort((a, b) -> {
            int freqA = map.get(a);
            int freqB = map.get(b);

            if (freqA != freqB) {
                return Integer.compare(freqB, freqA);
            }

            return Integer.compare(a, b);
        });

        return result;
    }
}